openssl/examples/mk_certs.rs

```diff
@@ -0,0 +1,143 @@
+use openssl::x509::{X509NameBuilder, X509Req, X509ReqBuilder};
+use openssl::x509::extension::{AuthorityKeyIdentifier, BasicConstraints, KeyUsage,
+                               SubjectAlternativeName, SubjectKeyIdentifier};
+
+/// Make a CA certificate and private key
+fn mk_ca_cert() -> Result<(X509, PKey), ErrorStack> {
+    let rsa = Rsa::generate(2048)?;
+    let privkey = PKey::from_rsa(rsa)?;
+
+    let mut x509_name = X509NameBuilder::new()?;
+    x509_name.append_entry_by_text("C", "US")?;
+    x509_name.append_entry_by_text("ST", "TX")?;
+    x509_name.append_entry_by_text("O", "Some CA organization")?;
+    x509_name.append_entry_by_text("CN", "ca test")?;
+    let x509_name = x509_name.build();
+
+    let mut cert_builder = X509::builder()?;
+    cert_builder.set_version(2)?;
+    let serial_number = ::openssl::bn::BigNum::from_dec_str("123")?;
+    cert_builder.set_serial_number(&&serial_number.to_asn1_integer()?)?;
+    cert_builder.set_subject_name(&x509_name)?;
+    cert_builder.set_issuer_name(&x509_name)?;
+    cert_builder.set_pubkey(&privkey)?;
+    let not_before = Asn1Time::days_from_now(0)?;
+    cert_builder.set_not_before(&not_before)?;
+    let not_after = Asn1Time::days_from_now(365)?;
+    cert_builder.set_not_after(&not_after)?;
+
+    cert_builder.append_extension(BasicConstraints::new().critical().ca().build()?)?;
+    cert_builder.append_extension(KeyUsage::new()
+        .critical()
+        .key_cert_sign()
+        .crl_sign()
+        .build()?)?;
+
+    let subject_key_identifier =
+        SubjectKeyIdentifier::new().build(&cert_builder.x509v3_context(None, None))?;
+    cert_builder.append_extension(subject_key_identifier)?;
+
+    cert_builder.sign(&privkey, MessageDigest::sha256())?;
+    let cert = cert_builder.build();
+
+    Ok((cert, privkey))
+}
+
+/// Make a X509 request with the given number of bits and the given common name
+fn mk_request(privkey: &PKey) -> Result<(X509Req), ErrorStack> {
+    let mut req_builder = X509ReqBuilder::new()?;
+    req_builder.set_pubkey(&privkey)?;
+
+    let mut x509_name = X509NameBuilder::new()?;
+    x509_name.append_entry_by_text("C", "US")?;
+    x509_name.append_entry_by_text("ST", "TX")?;
+    x509_name.append_entry_by_text("O", "Some organization")?;
+    x509_name.append_entry_by_text("CN", "www.example.com")?;
+    let x509_name = x509_name.build();
+    req_builder.set_subject_name(&x509_name)?;
+
+    req_builder.sign(&privkey, MessageDigest::sha256())?;
+    let req = req_builder.build();
+    Ok(req)
+}
+
+/// Make a certificate and private key signed by the given CA cert and private key
+fn mk_ca_signed_cert(ca_cert: &X509Ref, ca_privkey: &PKeyRef) -> Result<(X509, PKey), ErrorStack> {
+    let rsa = Rsa::generate(2048)?;
+    let privkey = PKey::from_rsa(rsa)?;
+
+    let req = mk_request(&privkey)?;
+
+    let mut cert_builder = X509::builder()?;
+    cert_builder.set_version(2)?;
+    let serial_number = ::openssl::bn::BigNum::from_dec_str("456")?;
```

Comment:
It seems a bit iffy to be hardcoding the serial number since it has to be unique per certificate per issuer.